petastorm/unischema.py

```diff
@@ -269,24 +287,69 @@ def as_spark_schema(self):
+        if "_fields" in dir(results):
+            field_names = results._fields
+        else:
+            field_names = results.keys()
+            if isinstance(field_names, KeysView):
+                # If received a dictionary, rather than a list of keys, grab keys from each item in the dictionary,
+                # then remove any duplicates.
+                field_names = []
+                for key in results.keys():
+                    field_names = field_names + list(results[key]._fields)
+                # Remove duplicates.
+                field_names = list(set(field_names))
+        return field_names
+
+    def verify_fields(self, results, expected_fields=None):
+        """ Confirm that actual data matches layout that we expected. If not, warn user about differences.
+        I.e. if data doesn't contain fields that we expect, or does have fields that didn't expect, emit warning.
+        """
+        if "_is_verified" not in self.__dict__:
+            # Can happen if read older version from a file.
+            self._is_verified = False
+
+        if not self._is_verified:
+            # Can only verify if have some data.
+            if results:
+                actual_field_names = self._get_field_names(results)
+                missing_fields = []
+                # See if there are fields in Unischema that we didn't receive in the returned data
+                for field in self._fields:
+                    if field not in actual_field_names:
+                        missing_fields.append(field)
+
+                # Then, see if there are any fields in the returned data that aren't in the schema.
+                unexpected_fields = []
+                for field in actual_field_names:
+                    if field not in self._fields:
+                        unexpected_fields.append(field)
+
+                # If found any mismatches, warn user.
+                msg = ""
+                if missing_fields:
+                    msg = msg + "The following field(s) were expected but not found in the returned data: " \
+                          + str(missing_fields) + "\n"
+                if unexpected_fields:
+                    msg = msg + "The following field(s) were found in the returned data, but were not expected: " \
+                          + str(unexpected_fields)
+                if msg:
+                    msghdr = "It seems that the transform function changed the fields that are returned, and these " \
+                             "changes weren't listed in the TransformSpec.edit_fields and removed_fields properties." \
+                             "If running under TensorFlow this will result in a crash, so please use these " \
+                             "parameters to avoid this crash. If not running under TensorFlow, please feel free to " \
+                             "ignore this warning.\n"
+                    warnings.warn(msghdr + msg)
+
+                self._is_verified = True
```

Comment:
Unclear why the state of verification of a particular `result` instance is stored in the schema? It could be that:
```python
some_schema.verify_fields(good_data) == True
# and then....
some_schema.verify_fields(bad_data) == True  # while should be False
```